mqtt/mqtt-bridge/src/pump/mod.rs

```diff
@@ -111,34 +120,106 @@ where
+                        }
+                    }
+                    Either::Right((ingress, egress)) => {
+                        if let Err(e) = ingress {
+                            debug!(error = %e, "ingress processing exited with error");
+                        } else {
+                            debug!("ingress processing exited");
+                        }
+
+                        if let Err(e) = egress.await {
+                            error!(error = %e, "egress processing exited with error");
+                        } else {
+                            info!("egress processing exited")
+                        }
+                    }
+                }
+
+                Ok(())
+            }
+            Either::Right((Either::Left((egress, ingress)), messages)) => {
+                if let Err(e) = &egress {
+                    debug!(error = %e, "egress processing exited with error");
+                } else {
+                    debug!("egress processing exited");
+                }
+
+                debug!("shutting down ingress...");
                 shutdown_ingress.shutdown().await;
+                if let Err(e) = ingress.await {
+                    error!(error = %e, "ingress processing exited with error");
+                } else {
+                    info!("ingress processing exited")
+                }
+
+                debug!("shutting down pump messages processor...");
+                shutdown_messages.shutdown().await;
+                if let Err(e) = messages.await {
+                    debug!(error = %e, "pump messages processor exited with error");
+                } else {
+                    debug!("pump messages processor exited");
+                }
+
+                egress.map_err(|e| PumpError::Run(e.into()))
+            }
+            Either::Right((Either::Right((ingress, egress)), messages)) => {
+                if let Err(e) = &ingress {
+                    debug!(error = %e, "ingress processing exited with error");
+                } else {
+                    debug!("ingress processing exited");
+                }
+
+                debug!("shutting down egress...");
                 shutdown_egress.shutdown().await;
+                if let Err(e) = egress.await {
+                    error!(error = %e, "egress processing exited with error");
+                } else {
+                    info!("egress processing exited")
+                }
 
-                join!(egress, ingress);
+                debug!("shutting down pump messages processor...");
+                shutdown_messages.shutdown().await;
+                if let Err(e) = messages.await {
+                    debug!(error = %e, "pump messages processor exited with error");
+                } else {
+                    debug!("pump messages processor exited");
+                }
+
+                ingress.map_err(|e| PumpError::Run(e.into()))
             }
-        };
+        }?;
```

Comment:
Seems a bit weird we only raise error in case first-exiting thing is an error. Shouldn't we log and raise in all error cases?

Reply:
good point